go/vt/vtgateproxy/discovery.go

```diff
@@ -48,225 +41,312 @@ var (
+
+	err := jsonDiscovery.start()
+	if err != nil {
+		return nil, err
 	}
 
-	log.V(100).Infof("Start discovery for target %v poolType %s affinity %s\n", target.URL.String(), poolType, affinity)
+	return jsonDiscovery, nil
+}
+
+func (*JSONGateResolverBuilder) Scheme() string { return "vtgate" }
 
-	r := &JSONGateConfigResolver{
-		target:   target,
-		cc:       cc,
-		jsonPath: b.JsonPath,
-		poolType: poolType,
-		affinity: affinity,
+// Parse and validate the format of the file and start watching for changes
+func (b *JSONGateResolverBuilder) start() error {
+
+	b.rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	// Perform the initial parse
+	_, err := b.parse()
+	if err != nil {
+		return err
 	}
-	r.start()
-	return r, nil
-}
-func (*JSONGateConfigDiscovery) Scheme() string { return "vtgate" }
 
-func RegisterJsonDiscovery() {
-	jsonDiscovery := &JSONGateConfigDiscovery{
-		JsonPath: *jsonDiscoveryConfig,
+	// Validate some stats
+	if len(b.targets) == 0 {
+		return fmt.Errorf("no valid targets in file %s", b.jsonPath)
 	}
-	resolver.Register(jsonDiscovery)
-	log.Infof("Registered JSON discovery scheme %v to watch: %v\n", jsonDiscovery.Scheme(), *jsonDiscoveryConfig)
-}
 
-// Resolver(https://godoc.org/google.golang.org/grpc/resolver#Resolver).
-type JSONGateConfigResolver struct {
-	target   resolver.Target
-	cc       resolver.ClientConn
-	jsonPath string
-	poolType string
-	affinity string
+	// Log some stats on startup
+	poolTypes := map[string]int{}
+	affinityTypes := map[string]int{}
 
-	ticker *time.Ticker
-	rand   *rand.Rand // safe for concurrent use.
-}
+	for _, t := range b.targets {
+		count := poolTypes[t.poolType]
+		poolTypes[t.poolType] = count + 1
 
-func min(a, b int) int {
-	if a < b {
-		return a
+		count = affinityTypes[t.affinity]
+		affinityTypes[t.affinity] = count + 1
 	}
-	return b
-}
 
-func jsonDump(data interface{}) string {
-	json, _ := json.Marshal(data)
-	return string(json)
-}
+	log.Infof("loaded %d targets, pool types %v, affinity groups %v", len(b.targets), poolTypes, affinityTypes)
 
-func (r *JSONGateConfigResolver) resolve() (*[]resolver.Address, []byte, error) {
+	// Start a config watcher
+	b.ticker = time.NewTicker(100 * time.Millisecond)
+	fileStat, err := os.Stat(b.jsonPath)
+	if err != nil {
+		return err
+	}
 
-	log.V(100).Infof("resolving target %s to %d connections\n", r.target.URL.String(), *numConnections)
+	go func() {
+		for range b.ticker.C {
+			checkFileStat, err := os.Stat(b.jsonPath)
+			if err != nil {
+				log.Errorf("Error stat'ing config %v\n", err)
+				continue
+			}
+			isUnchanged := checkFileStat.Size() == fileStat.Size() || checkFileStat.ModTime() == fileStat.ModTime()
+			if isUnchanged {
+				// no change
+				continue
+			}
 
-	data, err := os.ReadFile(r.jsonPath)
+			fileStat = checkFileStat
+
+			contentsChanged, err := b.parse()
+			if err != nil || !contentsChanged {
+				continue
+			}
+
+			// notify all the resolvers that the targets changed
```

Comment:
This is the core of the change - the resolver "builder" is responsible for watching the file.

Each target then gets a separate resolver class which we track in a list, and notify whenever something changes in the file, at which point the resolver turns around and picks a new set of hosts out of the global list.